src/core/fillOrder.ts

```diff
@@ -29,19 +35,100 @@ export const getTransactions = async (
+  // Default fill payload from the bundle
+  const originalFillPayload = {
     to: bundle.targetFillPayload.to,
     data: bundle.targetFillPayload.data,
     chainId: bundle.targetFillPayload.chainId,
     value: 0n, // we never need to send value
   }
 
+  // Apply inventory-based rebalancing (if relayer address is provided)
+  if (relayerAddress && bundle.acrossDepositEvents.length > 0) {
+    // Process each deposit event individually
+    let optimalChainId: number | null = null
+
+    // Loop through each deposit event to find the optimal destination
+    for (const depositEvent of bundle.acrossDepositEvents) {
+      try {
+        // Skip if there's no origin claim payload (already filtered in claims)
+        if (depositEvent.originClaimPayload.chainId === 0) continue
+
+        // Determine optimal destination chain for this deposit event
+        const eventOptimalChainId = await getOptimalDestinationChain(
+          depositEvent,
+          relayerAddress,
+          getRPCUrl,
+        )
+
+        // If we found a valid optimal chain, consider it
+        if (eventOptimalChainId) {
+          // If we haven't found an optimal chain yet, use this one
+          if (!optimalChainId) {
+            optimalChainId = eventOptimalChainId
+            debugLog(
+              `Considering chain ${optimalChainId} as optimal destination`,
+            )
+          }
+          // If this deposit's optimal chain differs from a previous one,
+          // we have conflicting needs. In this simple implementation,
+          // we'll use the first valid one, but a more sophisticated approach
+          // might prioritize based on deficit size, value, etc.
+          else if (optimalChainId !== eventOptimalChainId) {
+            debugLog(
+              `Note: Deposit events have different optimal destinations (${optimalChainId} vs ${eventOptimalChainId})`,
+            )
+          }
+        }
+      } catch (error) {
+        // If processing one deposit fails, log and continue with others
+        debugLog(`Error processing deposit event: ${error}`)
+        // We continue the loop to process other deposit events
+      }
+    }
+
+    // If we found an optimal chain different from the original destination
+    if (optimalChainId && optimalChainId !== bundle.targetFillPayload.chainId) {
+      debugLog(
+        `Rebalancing: Redirecting fill from chain ${bundle.targetFillPayload.chainId} to chain ${optimalChainId}`,
+      )
+
+      // TODO: In a real implementation, we would modify the fill transaction here
+      // to target the new chain. This would involve:
+      // 1. Generate a new transaction for the optimal chain
+      // 2. Update the fill payload with the new transaction
+
+      // For our stub implementation, we're just changing the chainId
+      // In production, this would require additional on-chain operations
+      const rebalancedFillPayload = {
+        ...originalFillPayload,
+        chainId: optimalChainId,
```

Comment:
this wont work - this tries to change the place where this claim/fill is made. instead to get repaid by across on a chain, you set the repaymentChainId. currently in the fill, this is arb by defaul but you would swap this out here. this is in the fill.data field which needs to be decoded, then the chainId swapped out and encoded again: https://github.com/across-protocol/contracts/blob/644239d408e7d132da4a47632567c052f03b0058/contracts/SpokePool.sol#L1026

Reply:
you're absolutely right. good catch, idk what I was thinking here, obviously we don't want to change the execution chain, we want to decode the calldata, change the repayment destination field, and re-encode.

Reply:
updated, please review! added some helpers to decode/modify/encode, I think I did it correctly?

I think this is an example of where hardhat testing might help, we can deploy with artifact added to artifacts directory (just added that for the SpokePool ABI), and actually test to make sure we're doing this correctly.